### ISSUE TITLE
How to integrate Swup in Wordpress?

### ISSUE THREAD

**Issue by ANONIMNIQ** [swup/swup#170]
Guys, I need help integrating swup to my wordpress site. I try everything, put `<main id="swup" class="transition-fade">` around content in my index.php, page.php, single.php, archive.php, search.php.. 

Create function in my function.php

```
 function pro_adding_scripts() {
	 wp_register_script('swup-pro-script', get_template_directory_uri() . '/swup.min.js', array('jquery'), '1.2', false);
	 wp_enqueue_script( 'swup-pro-script');
	 
	 wp_register_script('swup-pro-script2', get_template_directory_uri() . '/script.js', array('jquery'), '1.2', false);
	 wp_enqueue_script( 'swup-pro-script2');
}
add_action( 'wp_enqueue_scripts', 'pro_adding_scripts' ); 
```

import script in header.php

```
<script src="/swup.min.js"></script>
<script src="/script.js"></script>
```

Use this in my script.js file:

```
import Swup from 'swup';

const swup = new Swup({
 		LINK_SELECTOR: `a[href*="${domain}"]:not([data-no-swup]), a[href^="/"]:not([data-no-swup]), a[href^="#"]:not([data-no-swup]), a[xlink\\:href]`,
 	});
```

and add this css:

```
.transition-fade {
  transition: 0.4s;
  opacity: 1;
}

html.is-animating .transition-fade {
  opacity: 0;
}
```

Nothing happened... when press any link page just reloads normal. I'm not sure what I doing wrong.

**Comment by ANONIMNIQ** [comment]
I tried debug plugin,is what I see in the console when click a link (my domain name is replaced for security reasons)
```
GET http://mysite.com/smoothstate.min.js 404 (Not Found)
VM1509 swup.min.js:7 Uncaught SyntaxError: Unexpected token <
VM1511 script.js:1 Uncaught SyntaxError: Unexpected identifier
jquery-migrate.min.js?ver=1.4.1:2 JQMIGRATE: Migrate is installed, version 1.4.1
(index):97 GET http://mysite.com/wp-content/themes/mythemehttp:/mysite.com/swup.min.js?ver=1.2 net::ERR_ABORTED 404 (Not Found)
(index):98 GET http://mysite.com/wp-content/themes/mythemehttp:/mysite.com/script.js?ver=1.2 404 (Not Found)
swup.min.js:7 Uncaught SyntaxError: Unexpected token <
script.js:1 Uncaught SyntaxError: Unexpected identifier
(index):113 GET http://mysite.com/2017/12/21/time-to-forage/SwupDebugPlugin.js 404 (Not Found)
(index):211 Uncaught ReferenceError: anime is not defined
    at (index):211
```

**Comment by gmrchk** [comment]
Hi, 

this doesn't seem like a swup problem. Suggestions that could be a problem, but I'm not sure how accurate is your description:

---

You cannot use `import` in a JS file that goes straight to the browser. Imports need to be processed by bundlers that would make your code ready for a browser. This is actually mentioned in [docs](https://swup.js.org/getting-started/installation).  

You also shouldn't use `const` in a "straight to browser" code as well as the template literals (``). 

Also, I'm not really a WP guy so I don't know what the following code does, but swup doesn't have any external dependencies like jquery. 
```php
wp_register_script('swup-pro-script', get_template_directory_uri() . '/swup.min.js', array('jquery'), '1.2', false);
```

I actually think `wp_register_script` or `wp_enqueue_script` is the main problem here, as I don't see any reason why would  `swup.min.js` file throw an error in a first place. 

Path to SwupDebugPlugin also seems to be wrong. 

---

I will leave this open for a while so people can add their suggestions, but again, I don't think it's a swup issue. If you solve this, please, add a solution here.

**Comment by Nosrevelk** [comment]
Since you already have `<main id="swup" class="transition-fade">` and the css done. You just need to add the script tag and enable swup in the right location. 
If you use the header you will get the error:
`Cannot set property 'responseURL' of null`

Add the script before the closing body. (Edit the footer.php)

```
<script src="/swup.min.js"></script>
<script src="/script.js"></script>
</body>
</html>
```

Also change your script.js to only:
```
const swup = new Swup();
```

**Comment by gmrchk** [comment]
Closing this as it's an inactive thread. Feel free to open a new issue if you have some new info.

**Comment by daaanny90** [comment]
I have the same issue and the solution of @Nosrevelk, unfortunately, doesn't solve the problem. 
I get always ```swup.min.js:1 Uncaught TypeError: Cannot set property 'responseURL' of null``` also with the script in `footer.php` instead of `header.php`.
There is anybody that has successfully implemented Swup in WordPress?

**Comment by CFMazzarago** [comment]
Same problem of @daaanny90. I tried the solution of @Nosrevelk but doesn't work for me :( 
this let me crazy. Any suggestion?

**Comment by nateads** [comment]
@CFMazzarago I'm running it on a local build right now, hope the below info helps.
Inside functions.php where you enqueue your css and js files...
`wp_enqueue_script( 'swupjs', 'https://unpkg.com/swup@latest/dist/swup.min.js', array(), _S_VERSION, true);`

Also anything relating to swup goes into a separate js file named "swup.js"
`wp_enqueue_script( 'swup', get_template_directory_uri() . '/js/swup.js', array(), _S_VERSION, true);`

My swup.js file only contains...
`//Swup
const swup = new Swup( {
	plugins: [
		new SwupBodyClassPlugin(),
		new SwupScriptsPlugin(),
		new SwupScrollPlugin()
	],
});`

It's running for me with only a little bug of some of my carousel js isnt loading on page transitions, but it does load on refresh. But hopefully the above info will get you a running version of swup to work with.

**Comment by CFMazzarago** [comment]
hi @nateads and thanks! your suggestion helps me and now it works properly on wordpress. Thanks so much!

**Comment by QuirkyWallace** [comment]
How do you go about adding the plugins to wordpress? Is it the same as enqueuing the swup.js scripts?

**Comment by nateads** [comment]
@QuirkyWallace yes that is correct for my experience.

I dropped these lines into my functions.php inside the wp_enqueue_scripts function///
`wp_enqueue_script( 'nate-swupbodyjs', get_template_directory_uri() . '/js/SwupBodyClassPlugin.js', array(), _S_VERSION, true);

	wp_enqueue_script( 'nate-swupscriptsjs', get_template_directory_uri() . '/js/SwupScriptsPlugin.js', array(), _S_VERSION, true);

	wp_enqueue_script( 'nate-swupscrolljs', get_template_directory_uri() . '/js/SwupScrollPlugin.js', array(), _S_VERSION, true);`

Hope this info helps.

**Comment by QuirkyWallace** [comment]
Thank you @nateads. I used to get an error in the swup.js code for defining the plugins, but now all I get is `Cannot set property 'responseURL' of null`. I tried to define the container to an existing div in my wordpress theme as well as adding my own `div#swup`. No luck.

I tried to add the debug plugin but again, can't even get past that responseURL error.

I think I need to re-think where the container should be in the DOM.

**Comment by nateads** [comment]
@QuirkyWallace I'm using underscores.me and I placed #swup in my header.php as the first div right after <body> so it looks like this for me...
`<body <?php body_class(); ?>>
<?php wp_body_open(); ?>
<div id="swup" class="site transition-fade">`

I did it this way because I wanted as much of the page to change during a page transition.

Unfortunately I never ran into your error so I cannot pin point exactly where the issue is coming from.

Not sure if this post might help [https://github.com/swup/swup/issues/152](https://github.com/swup/swup/issues/152)

**Comment by QuirkyWallace** [comment]
I ended up starting over and taking your advice to place #swup as the first div after `<body>` and that error has been resolved!

Now I went to add the debug plugin by enqueuing it:
```
wp_enqueue_script('swupDebugPlugin', get_stylesheet_directory_uri() . '/assets/js/SwupDebugPlugin.min.js', array(), _S_VERSION, true);
```
And in my swup.js:
```
const swup = new Swup({
  plugins: [new SwupDebugPlugin()]
});
```
And I got this error:
```
Uncaught ReferenceError: SwupDebugPlugin is not defined
    at swup.js?ver=_S_VERSION:2
```

Perhaps I need to run the other plugins first?

**Comment by nateads** [comment]
@QuirkyWallace I believe that's correct but I'm not 100% sure. I think the debug should be last in line to load. Though I did run the debug plugin and found it wasn't helping me much. I found just trial and error worked better for debugging for me with swup.

**Comment by QuirkyWallace** [comment]
Hmm if I run Swup without plugins it can actually load the next page, but it's missing styles and inspecting the <body> I see those CSS classes are not updated. However I keep running into `SwupBodyClassPlugin is not defined` when trying to add the plugins.

I'll keep studying the documentation for what I should be doing. Thank you @nateads for the help!

**Comment by nateads** [comment]
@QuirkyWallace If it helps at all, here is my setup...
1. Ran npm to obtain plugin files I needed and moved the .js files over to my Wordpress /js/ folder.
2. Brought in the plugin files through wp_enqueue_scripts in my functions.php
3. Added the Swup ID and transition class to my first div in my header.php
4. Made a new js file specifically for swup called 'swup.js' and added it to my functions.php wp_enqeue_scripts
`wp_enqueue_script( 'nate-swup', get_template_directory_uri() . '/js/swup.js', array(), _S_VERSION, true);`
File contains...
`const swup = new Swup( {
	plugins: [
		new SwupBodyClassPlugin(),
		new SwupScriptsPlugin(),
		new SwupScrollPlugin()
	],
});

function init() {
	//my carousel jQuery
		}
	});

init();

swup.on('contentReplaced', init);`
5. Added css to my css file
`.transition-fade {
	transition: 0.4s;
	opacity: 1;
}

html.is-animating .transition-fade {
	opacity: 0;
}`
6. Added a filter for my wp_enqueue_scripts in functions.php
`add_filter( 'script_loader_tag', 'my_script_loader_tag', 10 ,2 );

function my_script_loader_tag( $tag, $handle ) {
	if ( $handle == 'nate-swup' ) {
		return str_replace( '<script', '<script data-swup-ignore-script', $tag );
	}
	return $tag;
}` this filter only loads the script once to avoid duplicate instances.

I think that's all I did, sorry it's been a few months since I did it. But that should be everything I think. Best of luck!

**Comment by QuirkyWallace** [comment]
Hello,

I resolved the plugin not defined error by **making sure I enqueue the plugin scripts before the Swup library and my JS script**
So the the order of scripts to be enqueued is:

1. Swup plugin scripts
2. Swup.js library
3. My Swup.js

I made sure the scripts are enqueued in this order by having [cascading dependencies](https://wordpress.stackexchange.com/questions/91961/hook-wp-enqueue-scripts-priority-has-no-effect) (second script has the first script as its dependency, third script has second, and so on and so on...) as well as giving priorities to my three enqueue functions.

Now all Swup scripts are enqueued and Swup is enabled on my [website](https://7ctestenv.site)! ...Although it looks like I still have to fix the theme loading styles after each page loads.

For anyone curious, I am using X Theme Pro and running a child theme.
Thanks for all the help @nateads!
